### ISSUE TITLE
Sik-Ho Tang | Review -- MoCo v2: Improved Baselines with Momentum Contrastive Learning.

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#142]
Sik-Ho Tang. [Review — MoCo v2: Improved Baselines with Momentum Contrastive Learning](https://sh-tsang.medium.com/review-moco-v2-improved-baselines-with-momentum-contrastive-learning-8b03598a0324).

**Comment by NorbertZheng** [comment]
## Overview
MoCo v2, Improves [MoCo](https://sh-tsang.medium.com/review-moco-momentum-contrast-for-unsupervised-visual-representation-learning-99b590c042a9) Using [SimCLR](https://sh-tsang.medium.com/review-simclr-a-simple-framework-for-contrastive-learning-of-visual-representations-5de42ba0bc66) Suggestions

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/dd3c2e89-4448-4d85-9aec-7e4ecfdf59e0)
A batching perspective of two optimization mechanisms for contrastive learning (Left: [SimCLR](https://sh-tsang.medium.com/review-simclr-a-simple-framework-for-contrastive-learning-of-visual-representations-5de42ba0bc66), Right: [MoCo](https://sh-tsang.medium.com/review-moco-momentum-contrast-for-unsupervised-visual-representation-learning-99b590c042a9)).

Improved Baselines with Momentum Contrastive Learning
MoCo v2, by Facebook AI Research (FAIR)
2020 arXiv, Over 700 Citations
Self-Supervised Learning, Unsupervised Learning, Contrastive Learning, Representation Learning, Image Classification, Object Detection

 - [SimCLR](https://sh-tsang.medium.com/review-simclr-a-simple-framework-for-contrastive-learning-of-visual-representations-5de42ba0bc66) systematically studies the major components for the success of self-supervised contrastive learning.
 - MoCo v2 is achieved by enhancing [MoCo](https://sh-tsang.medium.com/review-moco-momentum-contrast-for-unsupervised-visual-representation-learning-99b590c042a9) v1 based on [SimCLR](https://sh-tsang.medium.com/review-simclr-a-simple-framework-for-contrastive-learning-of-visual-representations-5de42ba0bc66)’s studies: using **an MLP projection head** and **more data augmentation**.
 - Indeed, this tech report only got 2 pages excluding the reference section.

**Comment by NorbertZheng** [comment]
## MoCo v2
### Background
 - Contrastive loss function, called InfoNCE, in [CPC](https://sh-tsang.medium.com/review-representation-learning-with-contrastive-predictive-coding-cpc-cpcv1-8ec7c6e91a8e), is used:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/b4df5ced-7b5e-4d6b-a06d-77565a1bdf4d)
where $q$ is a query representation, $k^{+}$ is a representation of the positive (similar) key sample, and $\\{k^{-}\\}$ are representations of the negative (dissimilar) key samples. $\tau$ is temperature hyper-parameter.

 - **End-to-End (Fig. a)**: In an end-to-end mechanism, the negative keys are from the same batch and updated end-to-end by back-propagation. [SimCLR](https://sh-tsang.medium.com/review-simclr-a-simple-framework-for-contrastive-learning-of-visual-representations-5de42ba0bc66) is based on this mechanism and requires a large batch to provide a large set of negatives.
 - **Memory Bank (Fig. b)**: The negative keys are maintained in a queue, and only the queries and positive keys are encoded in each training batch.
 - In [MoCo](https://sh-tsang.medium.com/review-moco-momentum-contrast-for-unsupervised-visual-representation-learning-99b590c042a9), a momentum encoder is adopted to improve the representation consistency between the current and earlier keys. [MoCo](https://sh-tsang.medium.com/review-moco-momentum-contrast-for-unsupervised-visual-representation-learning-99b590c042a9) decouples the batch size from the number of negatives.

**Comment by NorbertZheng** [comment]
### Improved Designs
 - [SimCLR](https://sh-tsang.medium.com/review-simclr-a-simple-framework-for-contrastive-learning-of-visual-representations-5de42ba0bc66) improves the end-to-end variant of instance discrimination in three aspects:
    - a substantially larger batch (4k or 8k) that can provide more negative sampples,
    - replacing the output fc projection head with an MLP head,
    - stronger data augmentation.
 - In the [MoCo](https://sh-tsang.medium.com/review-moco-momentum-contrast-for-unsupervised-visual-representation-learning-99b590c042a9) framework, a large number of negative samples are readily available.

>To enhance [MoCo](https://sh-tsang.medium.com/review-moco-momentum-contrast-for-unsupervised-visual-representation-learning-99b590c042a9), the MLP head and data augmentation are used.

**Comment by NorbertZheng** [comment]
## Experimental Results
### Temperature
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/dfecd597-2a15-4ec5-9bce-ad32da61fdbc)
Search for an optimal $\tau$ w.r.t. ImageNet linear classification accuracy.

 - Using the default $\tau=0.07$, pre-training with the MLP head improves from 60.6% to 62.9%; switching to the optimal value for MLP ($\tau=0.2$), the accuracy increases to 66.2%.

**Comment by NorbertZheng** [comment]
### MLP Head, Augmentation, Cosine Learning
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/6b23b730-594d-4cfe-9bd2-1e157ba1f65f)
Ablation of [MoCo](https://sh-tsang.medium.com/review-moco-momentum-contrast-for-unsupervised-visual-representation-learning-99b590c042a9) baselines, evaluated by ResNet-50 for (i) ImageNet linear classification, and (ii) fine-tuning VOC object detection (mean of 5 trials).

 - **MLP**: replace the fc head in [MoCo](https://sh-tsang.medium.com/review-moco-momentum-contrast-for-unsupervised-visual-representation-learning-99b590c042a9) with a 2-layer MLP head (**hidden layer (= 2048-d) 2048-d (> 128-d)**, with RELU).
 - **aug+**: Extra blur augmentation is added onto the default one.
 - **cos**: Cosine learning rate schedule.

>**MLP+aug+cos improves MoCo v1.** With longer training time of 800 epoches, results are even better.

**Comment by NorbertZheng** [comment]
### SOTA Comparison
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/b7ec60d9-8677-49fa-a284-507a8f55ca5f)
MoCo vs. [SimCLR](https://sh-tsang.medium.com/review-simclr-a-simple-framework-for-contrastive-learning-of-visual-representations-5de42ba0bc66): ImageNet linear classifier accuracy (ResNet-50, 1-crop 224×224).

>With MLP, aug+, cos, and longer training time (800 epoches), MoCo v2 achieves 71.1%, outperforming [SimCLR](https://sh-tsang.medium.com/review-simclr-a-simple-framework-for-contrastive-learning-of-visual-representations-5de42ba0bc66)’s 69.3% with 1000 epochs.

**Comment by NorbertZheng** [comment]
### Computational Cost
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/39c4fa3e-c00a-45ce-9c0f-53d55f30959b)
Memory and time cost in 8 V100 16G GPUs.

For end-to-end mechanism, the 4k batch size is intractable even in a high-end 8-GPU machine. Also, under the same batch size of 256, the end-to-end variant is still more costly in memory and time, because it back-propagates to both $q$ and $k$ encoders, while [MoCo](https://sh-tsang.medium.com/review-moco-momentum-contrast-for-unsupervised-visual-representation-learning-99b590c042a9) back-propagates to the $q$ encoder only.

**Comment by NorbertZheng** [comment]
Later on, [MoCo](https://sh-tsang.medium.com/review-moco-momentum-contrast-for-unsupervised-visual-representation-learning-99b590c042a9) extends as MoCo v3 by using [Vision Transformer, ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0).

**Comment by NorbertZheng** [comment]
## Reference
 - [2020 arXiv] [MoCo v2] [Improved Baselines with Momentum Contrastive Learning](https://arxiv.org/abs/2003.04297).
